File: forms/audiology-assessment/front-end-patient-form-with-svelte/src/lib/engine/hearing-grader.ts
import type { AssessmentData, HearingGrade, FiredRule } from './types';
import { hearingRules } from './hearing-rules';
import { classifyDbHL, worseGrade } from './utils';

/**
 * Pure function: evaluates all hearing rules against patient data.
 * Returns the maximum grade among all fired rules (worst finding),
 * defaulting to 'normal' for patients with no fired rules.
 *
 * The primary grading is based on the WHO Classification using
 * pure tone average (PTA) values, with additional rules for
 * speech audiometry, tinnitus, vestibular, and functional impact.
 */
export function calculateHearingGrade(data: AssessmentData): {
	hearingGrade: HearingGrade;
	firedRules: FiredRule[];
} {
	const firedRules: FiredRule[] = [];

	for (const rule of hearingRules) {
		try {
			if (rule.evaluate(data)) {
				firedRules.push({
					id: rule.id,
					system: rule.system,
					description: rule.description,
					grade: rule.grade
				});
			}
		} catch (e) {
			// Rule evaluation failed - log for debugging but continue grading
			console.warn(`Hearing rule ${rule.id} evaluation failed:`, e);
		}
	}

	// Determine grade from PTA values first (primary classification)
	const rightGrade = classifyDbHL(data.audiometricResults.pureToneAverageRight);
	const leftGrade = classifyDbHL(data.audiometricResults.pureToneAverageLeft);
	let ptaGrade = worseGrade(rightGrade, leftGrade);

	// Also consider fired rules that may elevate the grade
	const gradeOrder: Record<HearingGrade, number> = {
		normal: 0,
		mild: 1,
		moderate: 2,
		severe: 3,
		profound: 4
	};

	let hearingGrade: HearingGrade = ptaGrade;
	if (firedRules.length > 0) {
		const maxRuleGrade = firedRules.reduce<HearingGrade>((max, r) => {
			return gradeOrder[r.grade] > gradeOrder[max] ? r.grade : max;
		}, 'normal');
		hearingGrade = worseGrade(ptaGrade, maxRuleGrade);
	}

	return { hearingGrade, firedRules };
}

File: forms/audiology-assessment/full-stack-with-rust-axum-loco-tera-htmx-alpine/src/engine/hearing_grader.rs
use super::hearing_rules::all_rules;
use super::types::{AssessmentData, FiredRule, HearingLevel};
use super::utils::better_ear_pta;

/// Pure function: evaluates all hearing rules against assessment data.
/// Returns the hearing level, pure-tone average (dB HL), and fired rules.
/// If insufficient audiometric data is available, returns "draft" status.
pub fn calculate_hearing_level(data: &AssessmentData) -> (HearingLevel, f64, Vec<FiredRule>) {
    // Calculate the better-ear PTA
    let pta = match better_ear_pta(data) {
        Some(pta) => pta,
        None => return ("draft".to_string(), 0.0, vec![]),
    };

    // Fire rules
    let mut fired_rules = Vec::new();
    for rule in all_rules() {
        if (rule.evaluate)(data) {
            fired_rules.push(FiredRule {
                id: rule.id.to_string(),
                category: rule.category.to_string(),
                description: rule.description.to_string(),
                concern_level: rule.concern_level.to_string(),
            });
        }
    }

    // Determine hearing level from better-ear PTA (WHO classification)
    let level = if pta <= 25.0 {
        "normal"
    } else if pta <= 40.0 {
        "mild"
    } else if pta <= 55.0 {
        "moderate"
    } else if pta <= 70.0 {
        "moderatelySevere"
    } else if pta <= 90.0 {
        "severe"
    } else {
        "profound"
    };

    (level.to_string(), pta, fired_rules)
}
